### PR TITLE
Modify Trie structure

### PR DESCRIPTION
**Pull request by danilbeltsov**
-  Make Trie extends from Map
-  Redo methods
- Add test coverage

### CODE REVIEW

File: src/Trie.js [hunk 1]
```diff
@@ -9,31 +9,12 @@ const getClosestNode = require('./getClosestNode')
...
   @example const store = new Trie()
 */
-class Trie {
+class Trie extends Map {
   rootSocket: Object
-  cache: Map<string, any>
-  size: number
 
   constructor () {
+    super()
```
Comment (viktor-ku) [inline src/Trie.js]: Will this work?

```js
const a = new Map()
a.set('John Michael', { id: 1} )

const store = new Store(a)
store.match('John') // => { id: 1 }
```

File: test/has.js [hunk 2]
```diff
@@ -0,0 +1,20 @@
...
+const t = require('tap')
+const Store = require('../src/Trie')
+const $ = require('./lib/state')
+
+t.test('has', t => {
+  const store = new Store()
+  const name = 'Michael Jackson'
+
+  $.feed(store)
+  t.equal(store.has(name), true)
```
Comment (viktor-ku) [inline test/has.js]: Use just `t.ok` please

File: test/has.js [hunk 3]
```diff
@@ -0,0 +1,20 @@
...
+
+  store.delete(name)
+
+  t.equal(store.has(name), false)
```
Comment (viktor-ku) [inline test/has.js]: `t.notOk`

**Review by viktor-ku — CHANGES_REQUESTED** [review]

File: src/Trie.js [hunk 4]
```diff
@@ -7,33 +7,20 @@ const getClosestNode = require('./getClosestNode')
 /**
   Trie class
   @name Trie
-  @example const store = new Trie()
+  @example const store = new Trie(...args)
```
Comment (viktor-ku) [inline src/Trie.js]: Bad example. We don't do something like this in a real world

```js
const store = new Trie(...args)
```

Specify something useful instead

```js
const store = new Trie(new Map())
```

File: src/Trie.js [hunk 5]
```diff
@@ -7,33 +7,20 @@ const getClosestNode = require('./getClosestNode')
...
-  }
-
-  keys () {
-    // TODO
-  }
-
-  values () {
-    // TODO
+    if (args) {
+      args.forEach((value, key, table) => {
```
Comment (viktor-ku) [inline src/Trie.js]: We don't use `table` anywhere so why define it?

File: src/Trie.js [hunk 6]
```diff
@@ -7,33 +7,20 @@ const getClosestNode = require('./getClosestNode')
...
+        self.set(key, value)
+      })
+    }
   }
 
   [Symbol.iterator] () {
```
Comment (viktor-ku) [inline src/Trie.js]: Could you explain why do you ever need this? Isn't it something that you've already extended from the Map?

File: src/Trie.js [hunk 7]
```diff
@@ -136,7 +103,7 @@ class Trie {
     @example store.set('Anton Webern', [])
     @example store.set('Charles Best', function info () {})
   */
-  set (key: string, value: any): void {
+  set (key: string, value: any): Map {
```
Comment (viktor-ku) [inline src/Trie.js:100]: Do you really return `Map` from this method? Add test case for it please

File: src/Trie.js [hunk 8]
```diff
@@ -153,19 +120,19 @@ class Trie {
...
   }
 
   /**
-    Removes any value associated to the key and returns the value that has(key) would have previously returned. has(key) will return false afterwards
+    Delete any value associated to the key and returns the value that has(key) would have previously returned. has(key) will return false afterwards
 
-    @example store.remove('Michael Jacobs')
+    @example store.delete('Michael Jacobs')
   */
   delete (query: string): void {
```
Comment (viktor-ku) [inline src/Trie.js]: From MDN about `delete`:

> Returns true if an element in the Map object existed and has been removed, or false if the element does not exist.

Check that your methods does exactly that and test it

File: test/forEach.js [hunk 9]
```diff
@@ -10,55 +10,9 @@ t.test('forEach', t => {
   t.test('invoke callback for each iteration', t => {
```
Comment (viktor-ku) [inline test/forEach.js:10]: Do we ever test `forEach`?

**Review by viktor-ku — CHANGES_REQUESTED** [review]

File: src/Trie.js (hunk 4 above)
Comment (danilbeltsov) [reply src/Trie.js]: ok. understood

File: src/Trie.js (hunk 6 above)
Comment (danilbeltsov) [reply src/Trie.js]: it's becouse we have our implementation of `set` for our structure, and works different

File: src/Trie.js (hunk 5 above)
Comment (danilbeltsov) [reply src/Trie.js]: My bad. We don't need it here

File: test/forEach.js (hunk 9 above)
Comment (danilbeltsov) [reply test/forEach.js:10]: rather yes than no. We compare sinon.count and store.size, and it tell us that we had iterated through object. but we can add some callback that do someting, do we?

File: src/Trie.js (hunk 6 above)
Comment (viktor-ku) [reply src/Trie.js]:
```js
[Symbol.iterator] () {
```

I mean this snippet

Comment (danilbeltsov) [reply src/Trie.js]: ahh. yes. i've already have it

File: src/Trie.js [hunk 10]
```diff
@@ -95,15 +89,15 @@ class Trie extends Map {
...
+    Sets the value for the key in the Trie object. 
 
     @example store.set('Michael Jackson', { id: 1 })
     @example store.set('Lord Kelvin', 1824)
     @example store.set('William Osler', 'Scientist')
     @example store.set('Anton Webern', [])
     @example store.set('Charles Best', function info () {})
   */
-  set (key: string, value: any): Map {
+  set (key: string, value: any): void {
```
Comment (viktor-ku) [inline src/Trie.js]: About `set` return again

> Sets the value for the key in the Map object. Returns the Map object.

So the annotation was correct but the implementation isn't

File: test/set.js [hunk 11]
```diff
@@ -15,5 +15,11 @@ t.test('set', t => {
     t.end()
   })
 
+  t.test('Return nothing', t => {
+    const store = new Store()
+    t.notOk(store.set($.MichaelJackson, 1))
```
Comment (viktor-ku) [inline test/set.js]: See the comment about `set` returns somewhere

**Review by viktor-ku — CHANGES_REQUESTED** [review]

**Review by viktor-ku — APPROVED** [review]

**Pull request merged**